File: Android_Java/TabsMaster/app/src/main/java/com/example/sergio/tabsmaster/Adapters/PersonAdapter.java
package com.example.sergio.tabsmaster.Adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import com.example.sergio.tabsmaster.Models.Person;
import com.example.sergio.tabsmaster.R;
import com.squareup.picasso.Picasso;

import java.util.List;

public class PersonAdapter extends BaseAdapter {

    private Context context;
    private int layout;
    private List<Person> persons;


    public PersonAdapter(Context context, int layout, List<Person> persons) {
        this.context = context;
        this.layout = layout;
        this.persons = persons;
    }

    @Override
    public int getCount() {
        return persons.size();
    }

    @Override
    public Object getItem(int position) {
        return persons.get(position);
    }

    @Override
    public long getItemId(int id) {
        return id;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup viewGroup) {
        ViewHolder holder;

        if (convertView == null) {
            convertView = LayoutInflater.from(context).inflate(layout, null);
            holder = new ViewHolder();
            holder.name = (TextView) convertView.findViewById(R.id.textViewPersonName);
            holder.country = (TextView) convertView.findViewById(R.id.textViewPersonCountry);
            holder.image = (ImageView) convertView.findViewById(R.id.imageViewFlag);
            convertView.setTag(holder);
        } else {
            holder = (ViewHolder) convertView.getTag();
        }

        Person currentPerson = (Person) getItem(position);
        // Person currentPerson = persons.get(position);

        // holder.name.setText(currentPerson.getName());
        // holder.country.setText(currentPerson.getCountry().getName());

        holder.name.setText(currentPerson.getName());
        holder.country.setText(currentPerson.getCountry().getName());

        //Sacar la bandera del pais
        String url = currentPerson.getCountry().getFlagURL();
        Picasso.with(context).load(url).into(holder.image);

        return convertView;
    }

    static class ViewHolder {
        private TextView name;
        private TextView country;
        private ImageView image;
    }
}

File: Android_Kotlin/ActivitiesListViewSharedPrefences/app/src/main/java/com/example/sergio/activitieslistviewsharedprefences/adapters/PersonAdapter.kt
package com.example.sergio.activitieslistviewsharedprefences.adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.TextView
import com.example.sergio.activitieslistviewsharedprefences.models.Person
import kotlinx.android.synthetic.main.list_view_person.view.*

class PersonAdapter(val context: Context, val layout: Int, val list: List<Person>) : BaseAdapter() {

    private val mInflator: LayoutInflater = LayoutInflater.from(context)

    override fun getItem(position: Int): Any {
        return list[position]
    }

    //Es la posicion pasada como long. Casi no se usa
    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    override fun getCount(): Int {
        return list.size
    }

    //La logica del codigo va en este metodo
    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val view: View
        val vh: PersonViewHolder

        //ConvertView representa los items de la listView para poder hacer scroll guardando los valores que hayan sido llamados
        //Si son nulos, se llama al converView al no haber sido creados antes
        if(convertView == null) {
            view = mInflator.inflate(layout, parent, false)
            vh = PersonViewHolder(view)

            //tag para inyectar objetos al view
            view.tag = vh
        } else {
            view = convertView
            vh = view.tag as PersonViewHolder
        }

        val fullName = "${list[position].firstname}, ${list[position].lastname}"
        vh.fullName.text = fullName
        vh.age.text = list[position].age.toString()

        return view
    }

}

private class PersonViewHolder(view: View) {
    val fullName: TextView = view.textViewName
    val age: TextView = view.textViewAge
}